### ISSUE TITLE
Control points validation

### ISSUE THREAD

**Issue by RazrFalcon** [nical/lyon#347]
Hi! This is not a lyon bug, but rather a question.

Let's take such an SVG and try to render it:
```xml
<svg xmlns="http://www.w3.org/2000/svg" width="200" height="200">
    <path d="M 174 31 C 174 38 173.1 151 173 151 C 120 150 34 151 18 151"
          fill="none" stroke="green" stroke-width="20"/>
</svg>
```

![1525979207](https://user-images.githubusercontent.com/725494/39888865-cca9ad90-549e-11e8-8acb-ba7d82e3e96a.png)

As you can see it was rendered in a completely different ways. Mainly because one of the control points is too close to end point (38 **173.1** 151 **173**). This can produce bizarre results.

My question is: do you know about this problem and does lyon have some methods to fix this issue?

PS:  I have no idea where to ask such questions.

**Comment by nical** [comment]
Hi! The odd behavior looks like it is caused by the fact that there the control point that is very close to the end-point is in the continuation of the direction of the previous segment and then the curve takes a very sharp turn (due to the distance between the two being so small). My guess is that:

- QPainter snaps the control point to the endpoint. That way at the endpoint the two tangents are clearly perpendicular and it decides to generate the rounded corner.
- Chrome just considers the direction of the two tangents by looking at the control points and since they are aligned it doesn't do the rounded join. The second segment then immediately takes a sharp turn upwards but that's after the join.
- Cairo... I have no clue, but it's fun.

This is really exacerbated by the fact that the distance between the control point and endpoint is much smaller than the line width.

FWIW, lyon accidentally/naively does the same as chrome in the stroke tessellator. Not necessarily by choice but because it is the default behavior you get with a simple tessellator if you don't try to do anything special for this type of cases and use the control points to compute the tangent at the endpoints (which is mathematically correct but in this case the tangent changes direction *very* quickly after the endpoint so it's a bit odd from a design perspective).

I honestly don't know what the best thing to do is. The Chrome approach is probably strictly speaking the correct one if the rule is that line joins are strictly only generated at end-points. My guess is that QPainter's result is nicer in the sense that it's probably what the designer drawing these curves would expect (sharp turn (almost) at endpoint = round join). How to define the threshold for that is tricky, though.

If your goal is to make it so all your drawing backends behave the same way, one way could be define a heuristic to snap the control points that are very close to endpoints. How to define what is close enough is probably something that takes into account the distance, the line width and maybe other parameters.

All of this is guess work, it'd be interesting to looks at QPainter's code to see what makes them the way they do (in particular to see if they handle this on purpose or if it's a byproduct of how they draw).

**Comment by RazrFalcon** [comment]
Thanks for a detailed answer! I thought that there is maybe a default/defined behavior for such a case, but looks like I have to improvise. 

>distance between the control point and endpoint is much smaller than the line width

That's it! For now I've tried to use fixed coefficients, like 0.5/0.25 and they work, but I don't like magic numbers. The stroke width will fit much better.

As for produced results and linejoin, in the example about it's the default one - *miter*. So Chrome is kinda correct. Here is how it looks with *round*:
![1526028370](https://user-images.githubusercontent.com/725494/39915802-8be93f32-5511-11e8-9b76-3e00cf8277a6.png)

and *bevel*:
![1526028402](https://user-images.githubusercontent.com/725494/39915809-8f09227c-5511-11e8-84f6-ce74be9f79d4.png)

\* Inkscape uses Cairo
\* Have no idea what Batik is using

So I presume that Chrome's behavior is the correct one. And QPainter simply mirrors the second control point.

PS:  should test via Firefox too.

**Comment by nical** [comment]
Oh I was assuming round joins by the look of QPainter's output.
Firefox will probably give you different results depending on whether you test on windows with hardware accelerated paining (using D2D) or any other configuration (using Skia, which is what chrome uses).
